### DISCUSSION TITLE
[Proposal] Add `methodof(...)` syntax for getting MethodInfo with automatic support for assembly trimming

### DISCUSSION THREAD

**Discussion by brantburnett** [dotnet/csharplang#5166] [Language Ideas]
When working with a LINQ providers and extensions or other forms of reflection, it is very common to retrieve MethodInfo or PropertyInfo objects via reflection. This is often done statically, with syntax like the following:

```cs
private static readonly System.Reflection.MethodInfo StringCompare =
    typeof(string).GetMethod("Compare", new[] { typeof(string), typeof(string) });
```

This has the significant downside that accessing an incorrect method signature is only found at runtime and not compile-time. Also, the syntax has recently become more complicated with the advent of trimming binaries. In order to make this safe for trimming, it's important to add additional attributes (based on my understanding):

```cs
[System.Runtime.CompilerServices.DynamicDependency("Compare", typeof(string))]
private static readonly System.Reflection.MethodInfo StringCompare =
    typeof(string).GetMethod("Compare", new[] { typeof(string), typeof(string) });
```

If generics are involved, the syntax becomes even more complicated, i.e.:

```cs
[System.Runtime.CompilerServices.DynamicDependency("OrderBy`2", typeof(Queryable)]
private static readonly System.Reflection.MethodInfo OrderBy =
    typeof(Queryable).GetMethods().Single(p => p.Name == nameof(Queryable.OrderBy) && p.GetParameters().Length == 2);

// Perhaps more correctly, but rarely written this way due to complexity and performance
// Note: This could definitely be optimized
[System.Runtime.CompilerServices.DynamicDependency("OrderBy`2", typeof(Queryable)]
private static readonly System.Reflection.MethodInfo OrderBy =
    typeof(Queryable).GetMethods().Single(p => 
        p.Name == nameof(Queryable.OrderBy) &&
        p.IsGenericMethodDefinition &&
        p.GetParameters().Length == 2 &&
        p.GetGenericArguments().Length == 2 &&
        p.GetParameters()[0].ParameterType == typeof(IQueryable<>).MakeGenericType(p.GetGenericArguments()[0]) &&
        p.GetParameters()[1].ParameterType == typeof(Expression<>).MakeGenericType(typeof(Func<,>).MakeGenericType(p.GetGenericArguments()[0], p.GetGenericArguments()[1])));
```

Both of these limitations could be addressed using a keyword that acts as syntactic sugar for the pattern, working similarly to the `nameof` keyword:

```cs
private static readonly System.Reflection.MethodInfo OrderBy =
    methodof(Queryable.OrderBy<TSource, TKey>(IQueryable<T>, Expression<Func<TSource, TKey>>));
```

A similar `propertyof` keyword could be used to support loading `PropertyInfo`.

Some specific suggestions:

1. This pattern should be usable anywhere within the class/struct/record
2. If used in a field initializer, the DynamicDependency property is applied to the field
3. If used in a method the DynamicDependency property is applied to the method
4. If used in a property, the DynamicDependency property is applied to the getter or setter method
5. If used in a property initializer, the DynamicDependency property is applied to the backing field
6. Only accessible members should be allowed (i.e. you may only reference private members within the same class)
7. The proper BindingFlags should be used based on the member's accessibility and if it is a static member
8. When working with generics, specific names like `TSource` or `TKey` wouldn't matter as compared to the original method definition. They are merely used internally within the `methodof` clause to reference the positional generic type parameters.

This would pattern would:

1. Ensure that the method signature exists at compile time
2. Vastly reduce the complexity for the developer
3. Allow the compiler to use a more optimized form, such as avoiding LINQ and caching items like `GetParameters()` for reuse within the loop
4. Automatically apply required `DynamicDependency` attributes to support trimming, a detail probably forgotten by most developers of shared libraries

Discussion points:

1. Should usage outside of a static field initializer or static constructor cache the result in a backing static field transparently? Or is the developer expected to cache in a static field if required?
2. What about constructors? Should the return type be `MethodBase` instead of `MethodInfo` for future flexibility?
3. What about getting the backing `get_XXX` and `set_XXX` methods for properties?

**Comment by HaloFour** [comment]
See: https://github.com/dotnet/roslyn/issues/128

Also look at point 4 under https://github.com/dotnet/csharplang/blob/3c8559f186d4c5df5d1299b0eaa4e139ae130ab6/meetings/2014/LDM-2014-10-15.md#rationale

The problem becomes that an operator like `methodof` would require a new syntax in order to disambiguate overloads.  This was punted with `nameof` where it was decided that a method group would be sufficient and it wasn't worth the added complexity.

That said, I would _love_ a `methodof` operator in the language, and I think the more use cases we can demonstrate the better.  There are enough BCL/framework APIs that depend on some form of reflection that work somewhat poorly with C# today where `methodof` would make that a lot cleaner and safer, and we have lots of precedence of the usefulness of `typeof` which follows a near identical pattern.

**Reply by brantburnett** [reply]
> See: [dotnet/roslyn#128](https://github.com/dotnet/roslyn/issues/128)
> 
> Also look at point 4 under https://github.com/dotnet/csharplang/blob/3c8559f186d4c5df5d1299b0eaa4e139ae130ab6/meetings/2014/LDM-2014-10-15.md#rationale
> 
> The problem becomes that an operator like `methodof` would require a new syntax in order to disambiguate overloads. This was punted with `nameof` where it was decided that a method group would be sufficient and it wasn't worth the added complexity.
> 
> That said, I would _love_ a `methodof` operator in the language, and I think the more use cases we can demonstrate the better. There are enough BCL/framework APIs that depend on some form of reflection that work somewhat poorly with C# today where `methodof` would make that a lot cleaner and safer, and we have lots of precedence of the usefulness of `typeof` which follows a near identical pattern.

I can certainly see the point about not letting reflection leak into the language, and I recognize the complexities around generic signatures. My concern is that reflection is such a key part of .NET, and the addition of trimming has made it even more complicated than it was before. There's just so much room for mistakes and that complexity just keeps increasing.

This is particular problematic for library developers. With Blazor and AOT compilation and trimming, a library that appears to work just fine in all of your tests may not when a consumer tries to use it. And the consumer might not even know until after it's deployed and dies in production in some edge case. Problems where you can't "trust" NuGet packages to be compatible could make it harder for application authors to adopt these new technologies.

**Comment by JustNrik** [comment]
Have you tried:
```cs
private static readonly MethodInfo OrderBy = 
    new Func<IQueryable<TSource>, Expression<Func<TSource, TKey>>, IOrderedQueryable<TSource>>(Queryable.OrderBy).Method;
```

**Reply by brantburnett** [reply]
@JustNrik 

That's interesting, I've never seen that approach before. It does work for non-generics, or cases where the generic parameters are all known, such as the `string.Compare` example. But it doesn't work for the `OrderBy` example because that's operating with unknown TSource and TKey which would be resolved later with `OrderBy.MakeGenericMethod(typeof(X), typeof(Y))`;

**Reply by JustNrik** [reply]
I mean, if at some point you will be able to do `typeof(X), typeof(Y)` it means that the type is actually known and you can use this approach (perhaps with a generic static class that will serve as a generic cache), unless you meant `x.GetType(), y.GetType()`. If so, I think you really don't have another option than your current approach

**Reply by brantburnett** [reply]
The generic base implementation is often stored in a static field for performance reasons, and then narrowed later in a generic method call.

A somewhat contrived example, but the sort of thing you'd see with LINQ:

```cs
public static class MyQueryableExtensions
{
    [System.Runtime.CompilerServices.DynamicDependency("OrderBy`2", typeof(Queryable)]
    private static readonly System.Reflection.MethodInfo OrderBy =
        typeof(Queryable).GetMethods().Single(p => p.Name == nameof(Queryable.OrderBy) && p.GetParameters().Length == 2);

    public static IQueryable<TSource> SortAndReverse<TSource, TKey>(this IQueryable<TSource> source, Expression<Func<TSource, TKey>> keySelector)
    {
        var orderByMethod = OrderBy.MakeGenericMethod(typeof(TSource), typeof(TKey));

        var sorted = (IOrderedQueryable<TSource>) orderByMethod.Invoke(null, new object[] { source, keySelector });

        return sorted.Reverse();
    }
}
```

**Reply by brantburnett** [reply]
Actually, that's not a great example, Looking at the runtime, I see how they were able to use your approach with generics.

https://github.com/dotnet/runtime/blob/ebba1d4acb7abea5ba15e1f7f69d1d1311465d16/src/libraries/System.Linq.Queryable/src/System/Linq/CachedReflection.cs#L539

Though it still leaves me with some questions about trimming and attributes. I would have thought that this would remove the need for trimming attributes, but I still see them applied at the OrderBy call site, and oddly for Enumerable not Queryable. https://github.com/dotnet/runtime/blob/ebba1d4acb7abea5ba15e1f7f69d1d1311465d16/src/libraries/System.Linq.Queryable/src/System/Linq/Queryable.cs#L263

I wonder if there's anyone who can explain this?

**Reply by JustNrik** [reply]
I mean, creating a new delegate is just 60 bytes and 1 nanosecond, `MakeGenericMethod` itself is at least 500 times slower, and not even mentioning `MethodInfo.Invoke` which also allocates an array. So I'd personally go with a delegate or try to make some black magic with linq expression api to compile a sort of delegate that works for all types but I'm unsure atm how would I go with that

**Reply by brantburnett** [reply]
I decided to test your assertion with some (quicky) benchmarks, and it appears that using the delegate approach is faster (though not 500x, and it does incur a bit more allocation overhead).

```
BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19043
Intel Core i7-9750H CPU 2.60GHz, 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=6.0.100-preview.5.21302.13
  [Host]     : .NET Core 5.0.9 (CoreCLR 5.0.921.35908, CoreFX 5.0.921.35908), X64 RyuJIT
  Job-PYKBDS : .NET Framework 4.8 (4.8.4400.0), X64 RyuJIT
  Job-HBGCPU : .NET Core 5.0.9 (CoreCLR 5.0.921.35908, CoreFX 5.0.921.35908), X64 RyuJIT


|       Method |        Job |       Runtime |    Toolchain |       Mean |   Error |  StdDev | Ratio | Rank |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|------------- |----------- |-------------- |------------- |-----------:|--------:|--------:|------:|-----:|-------:|------:|------:|----------:|
| CachedStatic | Job-PYKBDS |    .NET 4.6.1 |       net461 | 1,151.3 ns | 7.15 ns | 6.69 ns |  1.79 |    4 | 0.0801 |     - |     - |     505 B |
|     Delegate | Job-PYKBDS |    .NET 4.6.1 |       net461 |   793.5 ns | 1.96 ns | 1.83 ns |  1.24 |    3 | 0.0858 |     - |     - |     546 B |
| CachedStatic | Job-HBGCPU | .NET Core 5.0 | netcoreapp50 |   641.9 ns | 2.78 ns | 2.60 ns |  1.00 |    2 | 0.0763 |     - |     - |     480 B |
|     Delegate | Job-HBGCPU | .NET Core 5.0 | netcoreapp50 |   487.4 ns | 3.02 ns | 2.82 ns |  0.76 |    1 | 0.0811 |     - |     - |     512 B |
```

This makes me wonder why is the internal .NET LINQ implementation uses cached statics with `MakeGenericMethod`. Is there some other reason that this benchmark doesn't show? Or is it just an oversight?

Here's my benchmark code:

```cs
[MemoryDiagnoser]
public class Test
{
    public static MethodInfo OrderBy =
        new Func<IQueryable<object>, Expression<Func<object, object>>, IOrderedQueryable<object>>(Queryable.OrderBy).Method
            .GetGenericMethodDefinition();

    [Benchmark(Baseline = true)]
    public MethodInfo CachedStatic()
    {
        return OrderBy.MakeGenericMethod(typeof(string), typeof(int));
    }

    [Benchmark]
    public MethodInfo Delegate()
    {
        return new Func<IQueryable<string>, Expression<Func<string, int>>, IOrderedQueryable<string>>(
            Queryable.OrderBy).Method;
    }
}
```

**Reply by JustNrik** [reply]
yea I was exaggerating, though I'm very surprised they aren't much different in terms of performance in .net 5, I really didn't expect that (I actually expected it to be actually twice as slow but it's been improved a lot)

**Comment by MillKaDe** [comment]
I usually try to avoid explicit reflection by something like this:

[SharpLab link](https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQAwAIDKAZAougXnSjgDpsBLAOwEdTcAPABwCcBTAZw4oHsqOA3MjRYASoWJlRbAGYAbNgGMALrypCkwgMzEY6AMLoA3snTpgbBQHMAhsrbpqyrMpb6AtkwBiVRegAUJBg2ADSSGABGAJQaZmamkgBsxDqYoqQAsmzKABY8wACSVDI86BlwcAFRxglmmK4e3r7oHIoyEkGk+jyeNuyxcWmZ2XmFxaXuFBKtMsO5+QNmUADs6JMDAL4JJMlQqelZ82MlZXB6/tUmSHHEAKwAPEFhT45UygB8LW0dcKhdPUw+mxFmI5qMiidJtM2mCFrViKt1gkttc4ts4Lt9rDjqUMvAqjVUXU8PRmOwuGp7vU3J4fIpPg4iIFfuhQuF0NFCJ9Ot1euwAmzoiChodweM1lMmWxSAAhfIAT1ZHCwJNF+X0NjkckYrE43D4UWxIJWEs26MxoLVOLKMHOl3hOFwpN1FL4jweLxeTneDIkzOCzxZnII3N+/z5Dn8gpiDoOI3yEImkoC0rlwEVNmVjuxGq1OvJ+qohqtxsRFDNRPQ5pSxAALGUbNQCVcbgA3PoSioScqVC4g9ssTt6IjlO39juTfEj/F9+EDiW27uL2eolEbIA===)

```c#
using System;
using SLE = System.Linq.Expressions;
using SR = System.Reflection;

public class C {
  delegate int StrCmpFnc (string a, string b);
    
  static public SR.MethodInfo M11 () {
    StrCmpFnc scf = string.Compare;
    SR.MethodInfo mi = scf.Method;
    return mi;
  }
  static public SR.MethodInfo M12 () {
    Func<string, string, int> scf = string.Compare;
    SR.MethodInfo mi = scf.Method;
    return mi;
  }
    
  static public SR.MethodInfo M21 () {
    SLE.Expression<StrCmpFnc> e = (string a, string b) => string.Compare (a, b);
    SR.MethodInfo mi = (e.Body as SLE.MethodCallExpression).Method;
    return mi;
  }
  static public SR.MethodInfo M22 () {
    SLE.Expression<Func<string, string, int>> e = (string a, string b) => string.Compare (a, b);
    SR.MethodInfo mi = (e.Body as SLE.MethodCallExpression).Method;
    return mi;
  }
    
  static public void Main () {
    var mi11 = M11 ();
    var mi12 = M12 ();
    var mi21 = M21 ();
    var mi22 = M22 ();
  }
}
```

With `SLE.Expression`, it should be possible to get `PropertyInfo` and `FieldInfo` as well ...

**Comment by MillKaDe** [comment]
Here is my older post about a similar topic: [link](https://github.com/dotnet/csharplang/discussions/712#discussioncomment-69791)
